crates/ruff/src/rules/pylint/rules/import_self.rs

```diff
@@ -0,0 +1,81 @@
+use std::path::Path;
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::helpers::identifier_range;
+use ruff_python_ast::source_code::Locator;
+use rustpython_parser::ast::{Stmt, StmtKind};
+
+#[violation]
+pub struct ImportSelf;
+
+impl Violation for ImportSelf {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Module import itself")
+    }
+}
+
+fn get_module_from_path(path: &Path) -> Option<String> {
+    Some(
+        path.iter()
+            .last()?
+            .to_str()?
+            .strip_suffix(".py")?
+            .to_string(),
+    )
+}
+
+fn check_import_self(stmt: &Stmt, path: &Path, module_path: Option<&Vec<String>>) -> bool {
+    match &stmt.node {
+        StmtKind::Import { names, .. } => {
+            if let Some(module_path) = module_path {
+                names
+                    .iter()
+                    .any(|name| name.node.name == module_path.join("."))
```

Comment:
Nit: We should move the `.join` call out of the loop to avoid allocating a new `String` over every name
```suggestion
				let joined_module_path = module_path.join(".");
                names
                    .iter()
                    .any(|name| name.node.name == joined_module_path)
```

Or we can use `split` on `name.node.name` to avoid allocating a new string entirely: 

```rust
  names
                    .iter()
                    .any(|name| name.node.name.split('.').eq(module_path.iter()))
```